v2/spanner-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/utils/SpannerToSourceDbExceptionClassifier.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (C) 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.google.cloud.teleport.v2.templates.utils;
+
+import com.datastax.oss.driver.api.core.type.codec.CodecNotFoundException;
+import com.google.cloud.spanner.ErrorCode;
+import com.google.cloud.spanner.SpannerException;
+import com.google.cloud.teleport.v2.spanner.exceptions.InvalidTransformationException;
+import com.google.cloud.teleport.v2.spanner.migrations.exceptions.ChangeEventConvertorException;
+import com.google.cloud.teleport.v2.templates.constants.Constants;
+import java.sql.SQLDataException;
+import java.sql.SQLNonTransientConnectionException;
+import java.sql.SQLSyntaxErrorException;
+import java.util.Set;
+import org.apache.beam.sdk.values.TupleTag;
+
+/** Helper class to classify Exceptions into Retryable or Permanent error. */
+public class SpannerToSourceDbExceptionClassifier {
+
+  private static final Set<ErrorCode> permanentSpannerErrorCodes =
+      Set.of(
+          ErrorCode.ALREADY_EXISTS,
+          ErrorCode.OUT_OF_RANGE,
+          ErrorCode.INVALID_ARGUMENT,
+          ErrorCode.NOT_FOUND,
+          ErrorCode.FAILED_PRECONDITION,
+          ErrorCode.PERMISSION_DENIED,
+          ErrorCode.UNAUTHENTICATED,
+          ErrorCode.RESOURCE_EXHAUSTED,
```

Comment:
RESOURCE_EXHAUSTED is supposed to be treated as a retryable error
"In general, if your application is experiencing RESOURCE_EXHAUSTED errors, treat the situation like an UNAVAILABLE error, and retry with exponential backoff." Full details - 
https://docs.cloud.google.com/spanner/docs/error-codes#error-codes